File: app/src/main/java/com/example/purepawapp/ui/admin/AdminPromotionAdapter.java
package com.example.purepawapp.ui.admin;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.purepawapp.data.model.Promotion;
import com.example.purepawapp.databinding.ItemAdminPromotionBinding;
import com.example.purepawapp.util.CurrencyUtils;
import com.example.purepawapp.util.DiscountType;

import java.text.SimpleDateFormat;
import java.util.Locale;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

public class AdminPromotionAdapter extends ListAdapter<Promotion, AdminPromotionAdapter.ViewHolder> {

    private final BiConsumer<Promotion, Boolean> onToggleActive;
    private final Consumer<Promotion> onDelete;

    private static final DiffUtil.ItemCallback<Promotion> DIFF_CALLBACK = new DiffUtil.ItemCallback<>() {
        @Override
        public boolean areItemsTheSame(@NonNull Promotion oldItem, @NonNull Promotion newItem) {
            return oldItem.getCode().equals(newItem.getCode());
        }

        @Override
        public boolean areContentsTheSame(@NonNull Promotion oldItem, @NonNull Promotion newItem) {
            return oldItem.isActive() == newItem.isActive() && oldItem.getUsedCount() == newItem.getUsedCount();
        }
    };

    public AdminPromotionAdapter(BiConsumer<Promotion, Boolean> onToggleActive, Consumer<Promotion> onDelete) {
        super(DIFF_CALLBACK);
        this.onToggleActive = onToggleActive;
        this.onDelete = onDelete;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemAdminPromotionBinding binding = ItemAdminPromotionBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private final ItemAdminPromotionBinding binding;

        ViewHolder(ItemAdminPromotionBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(Promotion promotion) {
            binding.tvCode.setText(promotion.getCode());
            binding.tvDescription.setText(promotion.getDescription());
            binding.tvValue.setText(DiscountType.PERCENTAGE.equals(promotion.getDiscountType())
                    ? "Giảm " + (int) promotion.getDiscountValue() + "%"
                    : "Giảm " + CurrencyUtils.toVndString(promotion.getDiscountValue()));
            SimpleDateFormat format = new SimpleDateFormat("dd/MM/yyyy", new Locale("vi", "VN"));
            binding.tvDateRange.setText(format.format(promotion.getStartDate()) + " - " + format.format(promotion.getEndDate()));

            int usageLimit = promotion.getUsageLimit();
            binding.progressUsage.setMax(100);
            binding.progressUsage.setProgress(usageLimit > 0
                    ? Math.min(Math.max(promotion.getUsedCount() * 100 / usageLimit, 0), 100)
                    : 0);
            binding.tvUsage.setText(usageLimit > 0
                    ? promotion.getUsedCount() + " / " + usageLimit + " lượt dùng"
                    : promotion.getUsedCount() + " lượt dùng · không giới hạn");

            binding.switchActive.setOnCheckedChangeListener(null);
            binding.switchActive.setChecked(promotion.isActive());
            binding.switchActive.setOnCheckedChangeListener((btn, isChecked) -> onToggleActive.accept(promotion, isChecked));
            binding.btnDelete.setOnClickListener(v -> onDelete.accept(promotion));
        }
    }
}

File: app/src/main/java/com/example/purepawapp/ui/admin/AdminPromotionAdapter.kt
package com.example.purepawapp.ui.admin

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.purepawapp.data.model.Promotion
import com.example.purepawapp.databinding.ItemAdminPromotionBinding
import com.example.purepawapp.util.DiscountType
import com.example.purepawapp.util.toVndString
import java.text.SimpleDateFormat
import java.util.Locale

class AdminPromotionAdapter(
    private val onToggleActive: (Promotion, Boolean) -> Unit,
    private val onDelete: (Promotion) -> Unit
) : ListAdapter<Promotion, AdminPromotionAdapter.ViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemAdminPromotionBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class ViewHolder(private val binding: ItemAdminPromotionBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(promotion: Promotion) {
            binding.tvCode.text = promotion.code
            binding.tvDescription.text = promotion.description
            binding.tvValue.text = if (promotion.discountType == DiscountType.PERCENTAGE) {
                "Giảm ${promotion.discountValue.toInt()}%"
            } else {
                "Giảm ${promotion.discountValue.toVndString()}"
            }
            val format = SimpleDateFormat("dd/MM/yyyy", Locale("vi", "VN"))
            binding.tvDateRange.text = "${format.format(promotion.startDate)} - ${format.format(promotion.endDate)}"

            val usageLimit = promotion.usageLimit
            binding.progressUsage.max = 100
            binding.progressUsage.progress = if (usageLimit > 0) {
                (promotion.usedCount * 100 / usageLimit).coerceIn(0, 100)
            } else {
                0
            }
            binding.tvUsage.text = if (usageLimit > 0) {
                "${promotion.usedCount} / $usageLimit lượt dùng"
            } else {
                "${promotion.usedCount} lượt dùng · không giới hạn"
            }

            binding.switchActive.setOnCheckedChangeListener(null)
            binding.switchActive.isChecked = promotion.isActive
            binding.switchActive.setOnCheckedChangeListener { _, isChecked -> onToggleActive(promotion, isChecked) }
            binding.btnDelete.setOnClickListener { onDelete(promotion) }
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Promotion>() {
            override fun areItemsTheSame(oldItem: Promotion, newItem: Promotion): Boolean = oldItem.code == newItem.code
            override fun areContentsTheSame(oldItem: Promotion, newItem: Promotion): Boolean = oldItem == newItem
        }
    }
}
